org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/LinkResourceUtil.java

```diff
@@ -0,0 +1,132 @@
+     * @param projectName name of the project.
+     * @param path path needs to be redirected.
+     * @return the redirected path.
+     */
+    private static IPath getMetaDataFilePath(String projectName, IPath path) {
+        if (path.segmentCount() == 1) {
+            return METADATA_FOLDER_PATH.append(projectName).append(path);
+        }
+
+        String lastSegment = path.lastSegment();
+        if (IProjectDescription.DESCRIPTION_FILE_NAME.equals(lastSegment)) {
+            return METADATA_FOLDER_PATH.append(projectName).append(lastSegment);
+        }
+
+        return null;
+    }
+
+	private static void linkFolderIfNewer(IFolder settingsFolder, Instant instant) throws CoreException {
+		if (settingsFolder.isLinked()) {
+			return;
+		}
+		if (settingsFolder.exists() && !isNewer(settingsFolder.getLocation().toPath(), instant)) {
+			return;
+		}
+		if (!settingsFolder.exists()) {
+			// not existing yet, create link
+			File diskFolder = getMetaDataFilePath(settingsFolder.getProject().getName(), settingsFolder.getProjectRelativePath()).toFile();
+			diskFolder.mkdirs();
+			settingsFolder.createLink(diskFolder.toURI(), IResource.NONE, new NullProgressMonitor());
+		} else if (isNewer(settingsFolder.getLocation().toPath(), instant)) {
+			// already existing but not existing before import: move then link
+			File sourceFolder = settingsFolder.getLocation().toFile();
+			File targetFolder = getMetaDataFilePath(settingsFolder.getProject().getName(), settingsFolder.getProjectRelativePath()).toFile();
+			File parentTargetFolder = targetFolder.getParentFile();
+			if (!parentTargetFolder.isDirectory()) {
+				parentTargetFolder.mkdirs();
+			}
+			sourceFolder.renameTo(targetFolder);
+			settingsFolder.createLink(targetFolder.toURI(), IResource.REPLACE, new NullProgressMonitor());
+		}
+	}
+
+	private static void linkFileIfNewer(IFile metadataFile, Instant instant, String ifEmpty) throws CoreException {
+		if (metadataFile.isLinked()) {
+			return;
+		}
+		if (metadataFile.exists() && !isNewer(metadataFile.getLocation().toPath(), instant)) {
+			return;
+		}
+		File targetFile = getMetaDataFilePath(metadataFile.getProject().getName(), metadataFile.getProjectRelativePath()).toFile();
+		targetFile.getParentFile().mkdirs();
+		try {
+			if (metadataFile.exists()) {
+				Files.move(metadataFile.getLocation().toFile().toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+			} else {
+				Files.writeString(targetFile.toPath(), ifEmpty != null ? ifEmpty : "", StandardOpenOption.CREATE);
+			}
+		} catch (IOException ex) {
+			throw new CoreException(Status.error(ex.getMessage(), ex));
+		}
+		metadataFile.createLink(targetFile.toURI(), IResource.REPLACE, new NullProgressMonitor());
+	}
+
+	public static void linkMetadataResourcesIfNewer(IProject project, Instant instant) throws CoreException {
+		linkFileIfNewer(project.getFile(IProjectDescription.DESCRIPTION_FILE_NAME), instant, null);
```

Comment:
Do we have any reason that always linking all of those meta data files, without checking its existence?

Reply:
Those files are very likely to be created at some point. By linking them early, then we know the link will be used and won't be created in the user folder.
If we don't link early, then files would be 1st created in user folder; and then we would have to move them afterwards.
Does this answer your question?

Reply:
Yes, it makes sense. Doing it early makes things easier afterwards :)